### ISSUE TITLE
Function .length mismatch when run from source code and load from bytecode on duktape 2.1 release

### ISSUE THREAD

**Issue by topelinux** [svaarala/duktape#1513]
Sample code
```javascript
var func = function (__filename, __dirname, module, exports, require) {
        function hello(a, b, c, d) {
            console.log('hello lengh in hello', hello.length);
        }

        console.log('hello.lengh', hello.length);
        hello();
};

func();
```
Run directly with `duk`

```
hello.lengh 4
hello lengh in hello 4
```

Compile to byte with `duk -c` and run again, the output is

```
hello.lengh 0
hello lengh in hello 0
```

**Comment by svaarala** [comment]
Thanks, I'll investigate.

**Comment by svaarala** [comment]
Hmm, I don't seem to be able to reproduce:

```
duktape@keppihevonen:~/duktape$ ./duk
((o) Duktape [linenoise] 2.1.0 (v2.1.0)
duk> Duktape.version
= 20100
duk> 
Cleaning up...

duktape@keppihevonen:~/duktape$ ./duk /tmp/test.js
hello.lengh 4
hello lengh in hello 4

duktape@keppihevonen:~/duktape$ ./duk -c /tmp/test.bc /tmp/test.js

duktape@keppihevonen:~/duktape$ ./duk /tmp/test.bc
hello.lengh 4
hello lengh in hello 4
```

**Comment by topelinux** [comment]
host: Macos
Darwin tope 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64 x86_64

reproduce steps:

* git clone https://github.com/svaarala/duktape-releases.git
* cd duktape-releases
* git checkout v2.1.0
* make -f Makefile.cmdline

```shell
gcc -o duk  -Os -pedantic -std=c99 -Wall -fstrict-aliasing -fomit-frame-pointer -I./src    -DDUK_CMDLINE_PRINTALERT_SUPPORT -I./extras/print-alert -DDUK_CMDLINE_CONSOLE_SUPPORT -I./extras/console -DDUK_CMDLINE_LOGGING_SUPPORT -I./extras/logging -DDUK_CMDLINE_MODULE_SUPPORT -I./extras/module-duktape src/duktape.c examples/cmdline/duk_cmdline.c extras/print-alert/duk_print_alert.c extras/console/duk_console.c extras/logging/duk_logging.c extras/module-duktape/duk_module_duktape.c -lm
```
* cat /tmp/test4.js
```
var func = function (__filename, __dirname, module, exports, require) {
        function hello(a, b, c, d) {
            console.log('hello lengh in hello', hello.length);
        }

        console.log('hello.lengh', hello.length);
        hello();
};

func();
```
* ./duk /tmp/test4.js

```
hello.lengh 4
hello lengh in hello 4
```

* ./duk -c test.bin /tmp/test4.js
* ./duk test.bin

```
hello.lengh 0
hello lengh in hello 0
```

**Comment by svaarala** [comment]
Thanks, I'll try those exact steps.

**Comment by svaarala** [comment]
Ok, I could reproduce it on Linux. Stay tuned.

**Comment by svaarala** [comment]
The bug seems to be in dumping inner function templates. Here the function instance is the entire JS file, while the function assigned to 'var func' and its inner function are both serialized as templates rather than function instances.

The concrete `.length` property was omitted from function templates to reduce their size. It is automatically defaulted from the internal `_Formals` property (which stores the formal argument list) which is serialized. That too is omitted if the function `nargs` matches `_Formals` length and this causes the problem.

The problem is not visible in the repo 'make duk' target because it enables debugger support, and when debugger support is enabled, `_Formals` is never dropped because it supports debugging.

This should be relatively straightforward to fix.

Thanks for reporting! :)

**Comment by svaarala** [comment]
To support the above theory: the `_Formals` property is kept also when a function contains an eval() call. If one adds an `eval("1")` to the innermost function in the example, the length fields come out correctly because it forces `_Formals` to be kept.

**Comment by svaarala** [comment]
Ok, #1516 will fix the issue and will be backported to 2.1.1 maintenance release. I don't have time to make the fix right now, but I'll do it in a few days.

**Comment by svaarala** [comment]
@topelinux #1516 should fix this issue, at least for the repro steps discussed above. If you encountered this in some non-trivial actual code, it'd be nice if you could check out #1516 to see if it works there too. Please reopen if you find any issues.

The 2.1.1 backport will need some thinking because it would need to be done without any bytecode dump format changes which are not allowed for patch versions. This may be challenging, but I'll see if that's possible.
